### PR TITLE
WebActionMetadata Endpoint

### PR DESCRIPTION
**Pull request by adrw**
* New endpoint that returns metadata for bound webactions:

```Kotlin
  name: String,
  function: Function<*>,
  functionAnnotations: List<Annotation>,
  acceptedMediaRanges: List<MediaRange>,
  responseContentType: MediaType?,
  parameterTypes: List<KType>,
  returnType: KType,
  pathPattern: PathPattern,
  applicationInterceptors: List<ApplicationInterceptor>,
  networkInterceptors: List<NetworkInterceptor>,
  dispatchMechanism: DispatchMechanism
```
* Follow up WebActionMetadata Tab coming soon...

### CODE REVIEW

File: misk/src/main/kotlin/misk/web/actions/WebActionMetadata.kt [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+data class WebActionMetadata(
+  val name: String,
+  val function: String,
+  val functionAnnotations: List<String>,
+  val requestMediaTypes: List<String> = listOf(),
+  val responseMediaType: String?,
+  val parameterTypes: List<String> = listOf(),
+  val returnType: String,
+  val pathPattern: String,
+  val interceptorsApplication: List<String> = listOf(),
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/actions/WebActionMetadata.kt]: interceptorsApplication –> applicationInterceptors

File: misk/src/main/kotlin/misk/web/actions/WebActionMetadata.kt [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  val interceptorsNetwork: List<String> = listOf(),
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/actions/WebActionMetadata.kt]: same

File: misk/src/main/kotlin/misk/web/actions/WebActionMetadata.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  val dispatchMechanism: DispatchMechanism
+)
+
+internal fun WebActionMetadata(
+  name: String,
+  function: Function<*>,
+  functionAnnotations: List<Annotation> = listOf(),
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/actions/WebActionMetadata.kt]: worth having defaults for these? Are they ever used?

File: misk/src/main/kotlin/misk/web/actions/WebActionMetadataAction.kt [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import javax.inject.Singleton
+
+@Singleton
+class WebActionMetadataAction : WebAction {
+  @Inject internal lateinit var servlet: WebActionsServlet
+
+  @Get("/api/webactionmetadata")
+  @RequestContentType(MediaTypes.APPLICATION_JSON)
+  @ResponseContentType(MediaTypes.APPLICATION_JSON)
+  @Unauthenticated
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/actions/WebActionMetadataAction.kt]: we’ll need to fix these ASAP

File: misk/src/main/kotlin/misk/web/actions/WebActionMetadataAction.kt [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  @Suppress("UNUSED_PARAMETER")
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/actions/WebActionMetadataAction.kt]: ?

File: misk/src/main/kotlin/misk/web/actions/WebActionMetadataAction.kt [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+import misk.web.RequestContentType
+import misk.web.ResponseContentType
+import misk.web.jetty.WebActionsServlet
+import misk.web.mediatype.MediaTypes
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class WebActionMetadataAction : WebAction {
+  @Inject internal lateinit var servlet: WebActionsServlet
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/actions/WebActionMetadataAction.kt]: mind injecting this as a `Provider<WebActionsServlet>` ? As-is there’s a little dependency cycle and this seems like the most natural place to break it.

```
@Inject internal lateinit var servletProvider: Provider<WebActionsServlet>
```

File: misk/src/main/kotlin/misk/web/jetty/WebActionsServlet.kt [hunk 7]
```diff
@@ -102,6 +105,26 @@ internal class WebActionsServlet @Inject constructor(
       realWebSocket.adapter
     }
   }
+
+  internal fun buildWebActionMetadata(): MutableList<WebActionMetadata> {
+    var webActionMetadata: MutableList<WebActionMetadata> = mutableListOf()
+    boundActions.map {
+      webActionMetadata.add(WebActionMetadata(
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/jetty/WebActionsServlet.kt]: should this be an extension on `BoundAction`, like so?

```
val BoundWebAction.metadata = ...
```

File: misk/src/main/kotlin/misk/web/actions/WebActionMetadata.kt (hunk 3 above)
Comment (adrw) [reply misk/src/main/kotlin/misk/web/actions/WebActionMetadata.kt]: `name` will always be used and can always be deduced by the action name which is a required field
`function` same for function
`functionAnnotations` might be used to discern roles/services but can be empty so `listOf()` seems to be a useful default

File: misk/src/main/kotlin/misk/web/actions/WebActionMetadataAction.kt (hunk 4 above)
Comment (adrw) [reply misk/src/main/kotlin/misk/web/actions/WebActionMetadataAction.kt]: to add an annotation for admin tab permissions?

File: misk/src/main/kotlin/misk/web/actions/WebActionMetadata.kt (hunk 3 above)
Comment (adrw) [reply misk/src/main/kotlin/misk/web/actions/WebActionMetadata.kt]: jk, I'll remove all the defaults

**Pull request merged**